### ISSUE TITLE
[BUG] Streaming requires Generator where Iterator should suffice

### ISSUE THREAD

**Issue by danielsc** [microsoft/promptflow#561]
**Describe the bug**
For the user code to provide streaming content, instead of the more generic Iterator protocol, Promptflow relies on the more specific `GeneratorType`. That is requiring the user to unnecessarily wrap their iterators in generators to stream content in more generic scenarios (such as streaming langchain results). Is there a reason PF depends on GeneratorType instead of the iterator protocol? Also, see this note in the [python source](https://github.com/python/cpython/blob/f65497fd252a4a4df960da04d68e8316b58624c0/Lib/types.py#L6-L10):
```
# Iterators in Python aren't a matter of type but of protocol.  A large
# and changing number of builtin types implement *some* flavor of
# iterator.  Don't check the type!  Use hasattr to check for both
# "__iter__" and "__next__" attributes instead.
```

**Concrete Issue:** 
When returning an iterator from my tool to enable streaming, I get the following error when running `pf flow test`
```
Flow test failed with UserErrorException: Exception: The output 'answer' for flow is incorrect. The output value is 
not JSON serializable. JSON dump failed: (TypeError) Object of type WordIterator is not JSON serializable. Please 
verify your flow output and make sure the value serializable.
```

**How To Reproduce the bug**
Here is my flow.dag.yaml:
```yaml
id: template_chat_flow
name: Template Chat Flow
inputs:
  chat_history:
    type: list
    default: []
  question:
    type: string
    default: What is the meaning of life?
outputs:
  answer:
    type: string
    is_chat_output: true
    reference: ${stream.output}
nodes:
- name: stream
  type: python
  source:
    type: code
    path: stream.py
  inputs:
    input: ${inputs.question}
```

And here is my stream.py
```python
from promptflow import tool

class WordIterator:
    def __init__(self, input: str):
        self.input = input

    def __iter__(self):
        return self

    def __next__(self):
        if self.input:
            word, *rest = self.input.split(" ")
            self.input = " ".join(rest)
            return f"{word} "
        else:
            raise StopIteration  

@tool
def my_python_tool(input: str):
    iterator = WordIterator(input)
    assert hasattr(iterator, "__iter__")
    assert hasattr(iterator, "__next__")
    return iterator
```
With the above PF run `pf flow test --flow .` -- I get this error:
```
2023-09-19 11:01:17 +0200   42558 execution          INFO     Start to run 1 nodes with concurrency level 16.
2023-09-19 11:01:17 +0200   42558 execution.flow     INFO     Executing node stream. node run id: c4ddaddd-6a38-44fd-9ab1-3258fb88bb37_stream_0
2023-09-19 11:01:17 +0200   42558 execution.flow     WARNING  Output of stream is not json serializable, use str to store it.
2023-09-19 11:01:17 +0200   42558 execution.flow     INFO     Node stream completes.
Flow test failed with UserErrorException: Exception: The output 'answer' for flow is incorrect. The output value is not JSON serializable. JSON dump failed: (TypeError) Object of type WordIterator is not JSON serializable. Please verify your flow output and make sure the value serializable.
```

**Expected behavior**
It should read out the iterator and stream the chunks to the caller -- `pf flow test --flow .` should look like this:
```
2023-09-19 10:34:05 +0200   40375 execution          INFO     Start to run 1 nodes with concurrency level 16.
2023-09-19 10:34:05 +0200   40375 execution.flow     INFO     Executing node stream. node run id: 24e60c4d-606a-4fc5-8e4c-cc4a5c41d6c8_stream_0
2023-09-19 10:34:05 +0200   40375 execution.flow     WARNING  Output of stream is not json serializable, use str to store it.
2023-09-19 10:34:05 +0200   40375 execution.flow     INFO     Node stream completes.
{
    "answer": "What is the meaning of life? "
}
```

**Running Information(please complete the following information):**
 - Promptflow Package Version using `pf -v`: 0.1.0b6
 - Operating System: 
```
ProductName:            macOS
ProductVersion:         13.5.2
BuildVersion:           22G91
```
 - Python Version using `python --version`: Python 3.10.12

**Additional context**
If I wrap the iterator in a generator, everything works as expected:

```python
from promptflow import tool

class WordIterator:
    def __init__(self, input: str):
        self.input = input

    def __iter__(self):
        return self

    def __next__(self):
        if self.input:
            word, *rest = self.input.split(" ")
            self.input = " ".join(rest)
            return f"{word} "
        else:
            raise StopIteration  
    
    def to_generator(self):
        try:
            while True:
                yield next(self)
        except StopIteration:
            pass

@tool
def my_python_tool(input: str):
    iterator = WordIterator(input).to_generator()
    assert hasattr(iterator, "__iter__")
    assert hasattr(iterator, "__next__")
    return iterator
```

**Comment by mindlessroman** [comment]
I'm seeing some similar behavior, slightly different execution.

# TLDR / Context

Using a node that uses the SDK for the AOAI tool returns what I expect, similar to how the examples/tutorials show. Using the package/tool in the `flow.dag.yaml` returns the `<generator object generate_from_proxy at MEMORY_LOCATION>`.

**THE ASK:** If something changed in how we should use the output of a test, can you point out where, or update the examples to show this? Runs seem to not be affected the same way.

## What works
`flow.dag.yaml`: Final step before output, that calls the llm
```yaml
inputs:
  question:
    type: string
outputs:
  answer:
    type: string
    reference: ${answer_the_question_with_context.output}
nodes:
...
- name: answer_the_question_with_context
  type: python
  source:
    type: code
    path: step05_answer_question_with_context.py
  inputs:
    prompt_text: ${STEP_FOUR.output}
    aoai_connection: some-aoai-connection
```

`step05_answer_question_with_context.py`: Connects to AOAI
```python
from promptflow import tool
from promptflow.connections import AzureOpenAIConnection
from promptflow.tools import AzureOpenAI
import os
from jinja2 import Template as jinja_template
@tool
def answer_question_with_context(
    prompt_text: jinja_template, 
    aoai_connection: AzureOpenAIConnection
) -> str:

    aoai_gpt_deployment = os.getenv("AZURE_OPENAI_GPT_DEPLOYMENT")

    response = AzureOpenAI(connection=aoai_connection).chat(
        prompt_text, 
        deployment_name=aoai_gpt_deployment, 
        temperature=0, 
        max_tokens=1000
    )

    return response
```

wil ultimately return: 

```python
test_result = pf.test(flow=FLOW_NAME, inputs=SOME_INPUTS)

print(f"Flow outputs: {test_result}")
```
```text
Flow outputs: {'answer': 'blah blah blah output that is relevant blah blah blah'}
```

## What *Doesn't* Work
`flow.dag.yaml` (again just showing the last step)
```yaml 
inputs:
  question:
    type: string
outputs:
  answer:
    type: string
    reference: ${answer_the_question_with_context.output}
nodes:
...
- name: answer_the_question_with_context
  type: llm
  source:
    type: code
    path: step05_answer_question_with_context.jinja2
  inputs:
    deployment_name: gpt-35-turbo
    temperature: 0
    max_tokens: 1000
    prompt_text: ${prompt_variants.output}
  connection: some-aoai-connection
  # provider: AzureOpenAI # have tried adding this
  api: chat
  # module: promptflow.tools.aoai # have tried adding this
```
`step05_answer_question_with_context.jinja2`: The prompt text generated in a previous step
```jinja
{{prompt_text}}
```

...all will ultimately return:

```python
test_result = pf.test(flow=FLOW_NAME, inputs=SOME_INPUTS)

print(f"Flow outputs: {test_result}")
```
```text
...
Flow outputs: {'answer': <generator object generate_from_proxy at MEMORY_LOCATION>}
```

and even with iterating with a for-loop:

```python
test_result = pf.test(flow=FLOW_NAME, inputs=SOME_INPUTS)

for result in test_result:
  print(f"Flow outputs: {result}")
```
```text
...
Flow outputs: answer
```
Where `answer` is whatever you call the output in the `flow.dag.yaml`.

**Comment by liucheng-ms** [comment]
Hi @mindlessroman,

Thank you for your feedback. I appreciate your effort in reporting this issue.

I understand that the issue you are facing is not related to the bug we are discussing in this thread. It seems to be a different problem.

From what I understand, you have a node that performs the LLM task. You encounter different results when you use different nodes:

- When you use a python node with a customized script, the flow works as expected.
- When you use an LLM node, the flow does not work as expected.

Is this correct?

I tried to reproduce the issue with the information you provided, but I could not replicate it on my side. Could you please open a new bug report and provide more details for us? This will help us investigate the issue further and find a solution.

Thank you for your cooperation.
